### PR TITLE
Add AMD to published packages

### PR DESCRIPTION
**Pull request by appsforartists**
The [Bazel build tool](https://github.com/bazelbuild/rules_typescript) currently requires dependencies to be in the AMD format.  This adds AMD to the list of things Rollup spits out.

There are a couple potential changes to be considered before merging:

1. **Should this be combined into the CommonJS export?**

   We could change the CommonJS export to use UMD instead, which would make that same build compatible with both CommonJS and AMD.

2. **How should `jsxstyle-utils` be packaged for the AMD bundle?**

   `jsxstyle-utils` is effectively an internal package; it gets included automatically for npm users.  However, AMD users wouldn't automatically receive it; they'd have to import both the `jsxstyle` and `jsxstyle-utils` bundles.  This seems less than ideal.

   **Here are two strategies for fixing that:**

   a. In `rollup.config.js`, abstract the config boilerplate into a function that decides whether or not to treat `jsxstyle-utils` as an `external` import.  Make it external for `cjs` and `esm`, but not for `amd`.

   b. Leave `rollup.config.js` as-is.  Add a `package.json` script that concatenates `jsxstyle-utils` onto each AMD bundle.

   Of course, both of these resolutions presume we distribute separate bundles for AMD and CommonJS.

Would love to hear your thoughts on these issues.

### CODE REVIEW

**Comment by meyer** [comment]
@appsforartists My currently uninformed and lightly-held opinion on the matter is that I’d prefer to not ship a UMD build. That said, the ES module build is what most build tools should be using, so the change from CJS to UMD would theoretically not affect most folks. I’ll look into building a UMD build instead of CommonJS and see what changes.

Regarding the `jsxstyle-utils` packaging issue, I’m going to need more context, as I have very limited knowledge of Bazel. For example, I don’t know why it wouldn’t be able to find a UMD build of `jsxstyle-utils`. Is there some documentation I can read? I looked at the bazel.build site and they don’t appear to mention TypeScript anywhere. Even better: do you have a sample project I could take a look at?

It seems like anyone who uses Bazel will run into issues with most popular modules. Do the Bazel folks have documentation on working with CommonJS modules? Or is the solution to convince module maintainers to switch from CommonJS to UMD 😜?

**Comment by appsforartists** [comment]
Bazel TypeScript is in development and doesn't have a good module story right now beyond AMD.  I expect they will eventually need to interoperate with ES modules, but haven't designed that feature yet.  I submitted this PR in-part because I want to be able to use `jsxstyle` from a project I'm experimenting with Bazel in, but it is unfortunately not yet public, so I can't share that repo.

AMD is a runtime dependency management format; it doesn't do static analysis/bundling.  This is how `jsxstyle` would be added to a `BUILD` file with AMD but not bundling:

```python
ts_devserver(
  name = "devserver",
  deps = [
   "//:node_modules/tslib/tslib.js",
    "//:node_modules/react/umd/react.development.js",
    "//:node_modules/react-dom/umd/react-dom.development.js",
    "//:node_modules/jsxstyle/jsxstyle.amd.js",
    "//:node_modules/jsxstyle-utils/jsxstyle-utils.amd.js",
    ":playground",
  ],
  entry_module = 'project_root/packages/playground/src/index',
  serving_path = "/dist/bundle.js",
  static_files = ["site/index.html"],
)
```

Notice that I had to know about `jsxstyle-utils` and import them.  Merely depending on another module from AMD doesn't automatically import it.  Concatenating the AMD builds together would resolve that issue.

Is your objection to including an AMD build in `dist`, or to converting the `CJS` one to `UMD`?

**Comment by meyer** [comment]
UMD in the place of CommonJS is… ok. It’s a bit verbose but bytes-wise it’s not a terrible addition, and it would enable people to drop a script tag in a page if they want to take `jsxstyle` for a spin.

Shipping a concatenated AMD build is definitely a no-go. Concatenating a build completely defeats the purpose of splitting `jsxstyle-utils` out as its own module. We’d basically be shipping a build just for Bazel (and arguably just for you). I could add an instruction portion to the README for Bazel support, but again, this is a one-off addition. Not ideal.

Seems to me like a better option would be for you to add a pre-build step to your project that would concat the latest version of `jsxstyle` and `jsxstyle-utils` and wrap the output in a UMD wrapper. Rollup would do the trick quite nicely—just whitelist `jsxstyle-utils` as an internal module and it’ll get somewhat intelligently concatenated (and tree shaken).

As far as an example goes, I don’t need to see your entire project. Just a representative idea (bazel config plus a single TS file, README would be a bonus) would do the trick.

It blows my mind that Bazel makes you manually specify every single dependency 🤯 Unless I’m misunderstanding something, this seems to completely defeat the purpose of modularising JS. I would have to assume that the folks working on Bazel’s TypeScript integration are well aware of how much this limitation sucks and that they’re actively working to support standard module formats.

---

To summarise:
1. I’m somewhat ok with the idea of UMD but I don’t like that it’s a change made for one very specific, very uncommon use case.
2. I’d rather add a “How to integrate with Bazel” section than provide a concatenated build, but since Bazel is so uncommon at this point in time, I’d rather not do either one.

I know this response probably seems a bit negative. I’m trying to understand the benefits this would bring to `jsxstyle` users.

**Comment by appsforartists** [comment]
To be honest, the only reason I care about AMD is because it's what Bazel uses - I've never built an AMD target before either.  

Since you have the build steps anyway, adding either an AMD build (without concatenating) or converting to UMD seems to be a reasonably small amount of work, but I can also understand the hesitation to do one.

I doubt there will be an AMD dependency in the long term in Bazel, but unfortunately, there is one now.

**Comment by meyer** [comment]
@appsforartists UMD build is definitely doable. Do you know what the standard method of distributing UMD modules is? I took a brief look at a few projects that distribute UMD modules, and it looks like building to a `umd` folder + adding `umd` to the `files` array in `package.json` is the day to go, but I’m not sure if that’s just for unpkg.com or if that’s the recommended method of distribution.

**Comment by appsforartists** [comment]
I don't know if there's a standard.  So long as it's distributed to npm, it's consumable from Bazel.

I imagine that browsers don't know about package.json and that bundlers already know how to parse CommonJS, so I don't think the AMD bit changes anything aside from a flag in Rollup, but I don't know much more than you do.

**Comment by meyer** [comment]
@appsforartists https://unpkg.com/jsxstyle/jsxstyle.amd.js 🎉

**Comment by appsforartists** [comment]
Sweeet!
